### PR TITLE
storage: write at provisional commit ts, not orig ts

### PR DESCRIPTION
**Pull request by benesch**
The fact that transactions write at their original timestamp, and not
their provisional commit timestamp, allows leaving an intent under a
read. The timestamp cache will ensure that the transaction can't
actually commit unless it can bump its intents above the read, but it
will still leave an intent under the read in the meantime.

This can lead to starvation. Intents are meant to function as a sort of
lock on a key. Once a writer lays down an intent, no readers should be
allowed to read above that intent until that intent is resolved.
Otherwise a continual stream of readers could prevent the writer from
ever managing to commit by continually bumping the timestamp cache.

Now consider how CDC's poller works: it reads (tsmin, ts1], then (ts1,
ts2], then (ts2, ts3], and so on, in a tight loop. Since it uses a
time-bound iterator under the hood, reading (ts2, ts3], for example,
cannot return an intent written at ts2. But the idea was that we
inductively guranteed that we never read above an intent. If an intent
was written at ts2, even though the read from (ts2, ts3] would fail to
observe it, the previous read from (ts1, ts2] would have.

Of course, since transactions write at their original timestamp, a
transaction with an original timestamp of ts2 can write an intent at ts2
*after* the CDC poller has read (ts1, ts2]. (The transaction will be
forced to commit at ts3 or later to be sequenced after the CDC poller's
read, but it will leave the intent at ts2.) The CDC poller's next read,
from (ts2, ts3], thus won't see the intent, nor will any future reads at
higher timestamps. And so the CDC poller will continually bump the
timestamp cache, completely starving the writer.

Fix the problem by writing at the transaction's provisional commit
timestamp (i.e., the timestamp that has been forwarded by the timestamp
cache, if necessary) instead of the transaction's original timestamp.
Writing at the original timestamp was only necessary to prevent a lost
update in snapshot isolation mode, which is no longer supported. In
serializable mode, the anomaly is protected against by the read refresh
mechanism.

Besides fixing the starvation problem, the new behavior is more
intuitive than the old behavior. It also might have some performance
benefits, as it is less likely that intents will need to be bumped at
commit time, which saves on RocksDB writes.

Touches #32433.

Release note: None

### CODE REVIEW

**Comment by tbg** [comment]
Are we exploring ripping out TBIs and pushing logic to C++? That sounded like the least scary venue.

**Comment by benesch** [comment]
> Are we exploring ripping out TBIs and pushing logic to C++? That sounded like the least scary venue.

My gut says that that won't be comparable to the benefit you get when you only need to scan the memtable and some L0 files. There is a fundamental limit to how fast scanning _n_ keys can be, and TBIs are currently our only way of reducing _n_ rather than the constant factor.

Perhaps we'd get lucky and things would be fast enough though. It's definitely worth an experiment.

**Comment by bdarnell** [comment]
> The CDC poller's next read,
from (ts2, ts3], thus won't see the intent, nor will any future reads at
higher timestamps. And so the CDC poller will continually bump the
timestamp cache, completely starving the writer.

Oh, this is subtle. The CDC poller doesn't actually care about uncommitted intents, and it *will* see the committed value at its new timestamp when and if it is able to commit. But the tight loop starves out the transaction so it can't commit. 

How does this change solve the problem? We write above the poller's read, but what stops the poller from advancing the timestamp cache again? Is it just that the poller stops advancing while there is a pending intent? 


We have some tests that use SNAPSHOT isolation; these tests will (hopefully) become flaky with a change like this. We should get rid of tests that use SNAPSHOT before we make changes that will break them. 

> But this change is definitely too scary to backport to 2.1.

We can't break SNAPSHOT in 2.1.x; clusters being upgraded from 2.0 may still perform snapshot-isolation transactions in a mixed-version configuration. 


If reads still use OrigTimestamp but writes use Timestamp, how does that affect commands that do both like ConditionalPut?

**Comment by benesch** [comment]
> Oh, this is subtle. The CDC poller doesn't actually care about uncommitted intents, and it will see the committed value at its new timestamp when and if it is able to commit. But the tight loop starves out the transaction so it can't commit.

I think it's actually possible that the CDC poller misses the value entirely. Say the poller reads at (ts1, ts2], and a concurrent transaction writes at ts1 (its orig timestamp) while getting pushed to ts3. If that transaction takes a while to send its commit RPC, the CDC poller might read (ts2, ts3] before the intent at ts3 gets resolved. And then it will be reading (ts3, ts4] by the time the intent gets resolved and never emit the value.

> How does this change solve the problem? We write above the poller's read, but what stops the poller from advancing the timestamp cache again? Is it just that the poller stops advancing while there is a pending intent?

Right. If the poller's ExportRequest catches a WriteIntentError, it won't update the timestamp cache, and so the intent can get bumped.

> We can't break SNAPSHOT in 2.1.x; clusters being upgraded from 2.0 may still perform snapshot-isolation transactions in a mixed-version configuration.

Good point.

> If reads still use OrigTimestamp but writes use Timestamp, how does that affect commands that do both like ConditionalPut?

I don't know. I suspect it works fine because all reads get refreshed to the commit timestamp anyway, so we're just doing extra work. But I need to think about this more.

File: pkg/roachpb/data.proto [hunk 1]
```diff
@@ -264,41 +264,7 @@ message Transaction {
...
-  // and commit, but this is not the expected outcome (the expected outcome is
-  // {key1: 6} (since txn1 is not expected to commit)) and we would be
-  // experiencing the Lost Update Anomaly.
-  //
-  // Note that in practice, before restarting, txn1 would still lay down an
-  // intent (just above the committed value) not with the intent to commit it,
-  // but to avoid being starved by short-lived transactions on that key which
-  // would otherwise not have to go through conflict resolution with txn1.
-  //
-  // Again, keep in mind that, when the transaction commits, all the intents are
```
Comment (andreimatei) [inline pkg/roachpb/data.proto]: this part of the comment should be kept somewhere. The comment on the `timestamp` field should be expended to explain how its used for writes, and how a txn might be laying down intents at different timestamps as it goes along.

File: pkg/storage/replica.go [hunk 2]
```diff
@@ -2746,6 +2746,7 @@ func (r *Replica) applyTimestampCache(
 						txn := ba.Txn.Clone()
 						bumped = txn.Timestamp.Forward(nextTS) || bumped
 						ba.Txn = &txn
+						ba.Timestamp = txn.Timestamp
```
Comment (andreimatei) [inline pkg/storage/replica.go]: please add a comment at this change and below, reiterating that we want to write at the bumped ts.

**Review by andreimatei — CHANGES_REQUESTED** [review]
LGTM mod the fact that this really needs tests for all the changes in `applyTimestampCache`.

I think we should couple this patch with a change to `PushTxn` which makes a push succeed if it's trying to push the txn at a ts <= its current timestamp (i.e. the timestamps known by the txn record). Like, if txn1 wrote at ts 5, and then it got bumped to ts 10, and now we have a read at ts 7, that read should succeed.
Otherwise, with this change, we can have situations where, because a txn now is leaving intents at different timestamps, whether or not two txn conflict can depend on which intent from the first the second one runs into.
In other words, I'm not so sure about the "the new behavior is more
intuitive than the old behavior" claim; the old one seems more intuitive to me. But otherwise I do think I like this change - there are benefits to writing at timestamps as high as possible.

**Comment by andreimatei** [comment]
Oh I wanted to say something else too - let's also couple this patch with deleting every trace of `snapshot` txns from our code. You can do it in a different commit if you want to backport the first (although the backport I guess would need to be put behind a cluster version per Ben's comments).

**Comment by bdarnell** [comment]
> although the backport I guess would need to be put behind a cluster version per Ben's comments

And backports aren't allowed to introduce cluster versions unless it's a really exceptional case.

**Comment by andreimatei** [comment]
> And backports aren't allowed to introduce cluster versions unless it's a really exceptional case.

We were just discussing offline how you don't really need a new cluster version, you could just gate on the `2.1` one.

**Review by benesch — COMMENTED** [review]
Ok, PTAL. I added a test and updated some more comments.

> LGTM mod the fact that this really needs tests for all the changes in applyTimestampCache.

Done.

> Oh I wanted to say something else too - let's also couple this patch with deleting every trace of snapshot txns from our code. You can do it in a different commit if you want to backport the first (although the backport I guess would need to be put behind a cluster version per Ben's comments)

Done in https://github.com/cockroachdb/cockroach/pull/32860.

> I think we should couple this patch with a change to PushTxn which makes a push succeed if it's trying to push the txn at a ts <= its current timestamp (i.e. the timestamps known by the txn record). Like, if txn1 wrote at ts 5, and then it got bumped to ts 10, and now we have a read at ts 7, that read should succeed.

Isn't this already the case? https://github.com/cockroachdb/cockroach/blob/2aef0420734b81b0bd042a5777e69c6b4fd226e1/pkg/storage/batcheval/cmd_push_txn.go#L161-L166

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-:-LTPIZZHBE8TrqjcB_ZP:b6aiplc)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 2749 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LRi2c515crZr8ehgUvX-r1-2749:-LTPIV0F3o181BkInFQD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7917d0f4c74c5517b2aba091662aa9f467e1582/pkg/storage/replica.go#L2749)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment at this change and below, reiterating that we want to write at the bumped ts.
</blockquote></details>

Done.

**Comment by benesch** [comment]
After lots of digging, my previous approach turned out to be a really bad idea. Bumping ba.Timestamp impacted both the read and the write timestamp. That meant a transaction could perform its reads at different timestamps and see an inconsistent snapshot of the database. This didn't cause correctness problems, since those transactions that saw inconsistent snapshots would fail to commit, but it made it impossible for clients to reason about the results of reads before the transaction committed, which confused the SQL layer's error handling.

Instead I've taken a more targeted approach of only bumping the write timestamp, without bumping the read timestamp. This was the original intent of the change, and seems to cause less problems. Looking for general feedback on this approach. I still have a comment or two to update, and a few test failures to fix.

Unlike last time, none of the test failures look particularly scary. A few places that used to get WriteTooOld errors now get TransactionRetry errors, and one very complicated merge test that depends on the implementation details of intent resolution now deadlocks.

**Comment by benesch** [comment]
I, naturally, spoke way too soon. There are some very scary failures indeed:

```
    backup_test.go:1717: error executing 'RESTORE data.* FROM $1': pq: importing 10 ranges: split at key /Table/54/1/900 failed: replica corruption (processed=true): range does not match splits: (/Table/54/1/600-/Table/54/1/900) + (/Table/54/1/900-/Max) != [n1,s1,r35/1:/Table/54/1/{600-700}]
```

I'll hopefully have time to investigate tomorrow.

**Comment by benesch** [comment]
It's greeeeen! 🍏📗💚🥗 Thank goodness.

**Review by bdarnell — APPROVED** [review]
> Reviewed 12 of 12 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32487)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/engine/mvcc.go, line 1310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1Xjb-1IeSI4VLfTNr:-LU1Xjb0C88RfDqeYV3z:b1vsrfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/engine/mvcc.go#L1310)):*
> ```Go
> 		// A txn writes intents at its provisional commit timestamp. See the
> 		// comment on the txn.Timestamp field definition for rationale.
> 		writeTimestamp.Forward(txn.Timestamp)
> ```

Should it be an error for `txn.Timestamp` to be less than `timestamp`? 
___
*[pkg/storage/engine/mvcc_test.go, line 436 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1Xo8J8rQXIj4h7Nnx:-LU1Xo8J8rQXIj4h7Nny:b3goyir) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/engine/mvcc_test.go#L436)):*
> ```Go
> 	txn := *txn1
> 	txn.Sequence++
> 	txn.Timestamp = hlc.Timestamp{WallTime: 1}
> ```

The tests would pass without this, right? (because of the way `.Forward` is used instead of just overwriting `timestamp` with `txn.Timestamp`)
___
*[pkg/storage/engine/mvcc_test.go, line 1114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1Xyzw-FQJyY7pfTno:-LU1Xyzw-FQJyY7pfTnp:b-udw10j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/engine/mvcc_test.go#L1114)):*
> ```Go
> 	ts := []hlc.Timestamp{{Logical: 1}, {Logical: 2}, {Logical: 3}, {Logical: 4}, {Logical: 5}, {Logical: 6}}
> 
> 	txn1ts := *txn1
> ```

Maybe the `txn1` global should be replaced with a factory function since it's no longer used as-is anywhere (or is it?).

**Review by andreimatei — APPROVED** [review]
LGTM mod a comment on the interface of MVCCPut.

And a question: What is the timestamp used for populating the write timestamp cache (by DelRange)?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-:-LU1xsJH87xUsLYEWHzs:bea9sh2)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica_test.go, line 2621 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1vaR847mI_jsd5buV:-LU1vaR847mI_jsd5buW:btnbuxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/replica_test.go#L2621)):*
> ```Go
> }
> 
> func TestReplicaTSCacheForwardsIntentTS(t *testing.T) {
> ```

pls add a comment to the test with some words
___
*[pkg/storage/replica_test.go, line 2666 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1wRbGDCMO-jm45lvt:-LU1wRbGDCMO-jm45lvu:b2mu3hx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/replica_test.go#L2666)):*
> ```Go
> 				t.Fatal(pErr)
> 			}
> 			iter := tc.engine.NewIterator(engine.IterOptions{Prefix: true})
> ```

respect for going medieval
___
*[pkg/storage/engine/mvcc.go, line 1310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1Xjb-1IeSI4VLfTNr:-LU1wsui3wImSH2m_O-z:bxkh25o) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/engine/mvcc.go#L1310)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should it be an error for `txn.Timestamp` to be less than `timestamp`? 
</blockquote></details>

yeah, I also wonder. And this function's interface, taking both a txn and a timestamp, is confusing. Please add some comments to these args. Maybe the `timestamp` can be documented and asserted to be nil if a `txn` is passed.
___
*[pkg/storage/engine/enginepb/mvcc3.proto, line 63 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1xoHq7fLtOGcG_Qu4:-LU1xoHq7fLtOGcG_Qu5:b7ip4zl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/engine/enginepb/mvcc3.proto#L63)):*
> ```protobuf
>   // original timestamp:
>   //
>   //    1. This timestamp is forwarded by the timestamp cache when this
> ```

good comment

**Review by benesch — COMMENTED** [review]
Whew, ok, addressed all the feedback. PTAL as the required changes to the storage/engine tests were rather large.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-:-LU49nu6137oynVYXSrY:bf0h699)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica_test.go, line 2621 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1vaR847mI_jsd5buV:-LU49IgC2xzWUR-bXggO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/replica_test.go#L2621)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls add a comment to the test with some words
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc.go, line 1310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1Xjb-1IeSI4VLfTNr:-LU47rgi5PQHUSzFBDq4:b-g6bseg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/engine/mvcc.go#L1310)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

yeah, I also wonder. And this function's interface, taking both a txn and a timestamp, is confusing. Please add some comments to these args. Maybe the `timestamp` can be documented and asserted to be nil if a `txn` is passed.
</blockquote></details>

Yes, it's extremely confusing. We can't use `nil` because hlc.Timestamp is passed by value. We don't really want to use `hlc.Timestamp{}` because that already means inline put. (Yes, technically you can't do an inline put within a txn, but seems bad to overload that sentinel.)

The best thing I came up with was to mandate that timestamp matches `txn.OrigTimestamp` (or `txn.RefreshedTimestamp` if that's set). This is quite redundant, but it matches today's usage and I can't come up with anything better that doesn't require a major refactor.

Note that enforcing this actually required a ton of changes to the tests in storage/engine, since those tests were frequently violating this new timestamp-must-be-the-txn's-read-timestamp rule.
___
*[pkg/storage/engine/mvcc_test.go, line 436 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1Xo8J8rQXIj4h7Nnx:-LU48__48AWg_TdHyPEi:b-dw7zz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/engine/mvcc_test.go#L436)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The tests would pass without this, right? (because of the way `.Forward` is used instead of just overwriting `timestamp` with `txn.Timestamp`)
</blockquote></details>

That's true in the revision you made this comment on, but is no longer true in the latest revision. See comment above for details.
___
*[pkg/storage/engine/mvcc_test.go, line 1114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1Xyzw-FQJyY7pfTno:-LU3hOOz9jum0ZL4-LP_:b-i5nl9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/engine/mvcc_test.go#L1114)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Maybe the `txn1` global should be replaced with a factory function since it's no longer used as-is anywhere (or is it?). 
</blockquote></details>

There's still a bunch of tests that use it directly. (Many tests are happy with txn1's default timestamp.)
___
*[pkg/storage/engine/enginepb/mvcc3.proto, line 63 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU1xoHq7fLtOGcG_Qu4:-LU3h0COCNYi7vVQdfgH:b75099u) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fddb13775f385a422f960ff9e10ceb01fb71558/pkg/storage/engine/enginepb/mvcc3.proto#L63)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

good comment
</blockquote></details>

Thank you! :)
___
*[pkg/storage/replica.go, line 2749 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LRi2c515crZr8ehgUvX-r1-2749:-LU3h5kG2o5UrInpF8Yg:b-cxchxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7917d0f4c74c5517b2aba091662aa9f467e1582/pkg/storage/replica.go#L2749)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

No longer relevant.

**Review by andreimatei — APPROVED** [review]
I was asking:
>  What is the timestamp used for populating the write timestamp cache (by DelRange)?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-:-LU6DF29ELwsTkG8Ht_B:bw44211)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/engine/mvcc.go, line 1267 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU6CEpuCvvUz3S_k9eZ:-LU6CEpv2_j5SFyBLOFT:bamtnsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1313362562a35d3ec57c8e46d09186790470b7aa/pkg/storage/engine/mvcc.go#L1267)):*
> ```Go
> // for inline puts.)
> //
> // TODO(benesch): explore alternate function signatures that reduce
> ```

umm... I don't think you should be putting your name in TODOs any more (unless you've reconsidered :P).

I'd encourage you to add an options struct to this method that lets one clearly express inline vs intent vs non-transactional, and boil the ocean :)

**Review by benesch — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32487)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/engine/mvcc.go, line 1267 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU6CEpuCvvUz3S_k9eZ:-LU7MYbd55viOFUmSYUO:b-f54oez) ([raw file](https://github.com/cockroachdb/cockroach/blob/1313362562a35d3ec57c8e46d09186790470b7aa/pkg/storage/engine/mvcc.go#L1267)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

umm... I don't think you should be putting your name in TODOs any more (unless you've reconsidered :P).

I'd encourage you to add an options struct to this method that lets one clearly express inline vs intent vs non-transactional, and boil the ocean :)
</blockquote></details>

Can I put your name on the TODO then? :sinceretroll:

I've got a bunch of merge stuff that's higher priority than this refactor, so I'm going to merge this as-is unless you object strongly. I think there's an 80% chance or so that I can actually get back to this refactor before I leave. (I've already done the refactor for MVCCGet (#33213) and MVCCScan (#32389).)

**Review by andreimatei — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32487)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/engine/mvcc.go, line 1267 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32487#-LU6CEpuCvvUz3S_k9eZ:-LU7RZVl08DaEU-5p6fX:bkwg08q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1313362562a35d3ec57c8e46d09186790470b7aa/pkg/storage/engine/mvcc.go#L1267)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Can I put your name on the TODO then? :sinceretroll:

I've got a bunch of merge stuff that's higher priority than this refactor, so I'm going to merge this as-is unless you object strongly. I think there's an 80% chance or so that I can actually get back to this refactor before I leave. (I've already done the refactor for MVCCGet (#33213) and MVCCScan (#32389).)
</blockquote></details>

yeah, use my name

**Comment by benesch** [comment]
🎉 🎉 🎉 

bors r=andreimatei,bdarnell
